### PR TITLE
[sqlite3] Adding support for SQLite options cross-compiling

### PR DESCRIPTION
**Pull request by AJIOB**
Adding possibility to specify target OS for compilation: windows-based, unix-based, other or auto-detected (that is used by default)

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/sqlite3/all/conanfile.py
Comment (SpaceIm) [inline recipes/sqlite3/all/conanfile.py:42]: It shouldn't be an option, but come from settings.

**Comment by SpaceIm** [comment]
In which circonstance does it fail? Cross compilation should work without manually injecting those options.

**Comment by AJIOB** [comment]
> 
> 
> In which circonstance does it fail? Cross compilation should work without manually injecting those options.

I'm trying to build sqlite with CMake & cross-compilation to baremetal, but I received of error including:
```
Default settings
	os=Linux
	os_build=Linux
	arch=x86_64
	arch_build=x86_64
	build_type=Release
*** You can change them in /builds/sqlite-build/conanCache/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***
Configuration:
[settings]
arch=armv7hf
arch_build=x86_64
build_type=Debug
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=7
os=Arduino
os.board=stm32h7-m4
os_build=Linux
[options]
sqlite3:build_executable=False
sqlite3:fPIC=False
sqlite3:shared=False
sqlite3:threadsafe=0
sqlite3:use_alloca=False
[build_requires]
[env]
AR=arm-none-eabi-ar
AS=arm-none-eabi-as
CC=arm-none-eabi-gcc
CHOST=arm-none-eabi
CONAN_CMAKE_TOOLCHAIN_FILE=/builds/sqlite-build/scripts/toolchains/stm32h7-m4.cmake
CXX=arm-none-eabi-g++
LD=arm-none-eabi-ld
RANLIB=arm-none-eabi-ranlib
STRIP=arm-none-eabi-strip
TOOLCHAIN_DIR=/opt/st/stm32cubeide_1.4.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.7-2018-q2-update.linux64_1.4.0.202007081208/tools
sqlite3/3.35.4: Not found in local cache, looking in remotes...
sqlite3/3.35.4: Trying with 'conan-center'...
Downloading conanmanifest.txt
Downloading conanfile.py
Downloading conan_export.tgz
sqlite3/3.35.4: Downloaded recipe revision 0
conanfile.py (SqliteBuild/build_lib_3ad5ded7): Installing package
Requirements
    sqlite3/3.35.4 from 'conan-center' - Downloaded
Packages
    sqlite3/3.35.4:70d57bfbf494755d82c86f4094add5b78d39e5d7 - Build
Cross-build from 'Linux:x86_64' to 'Arduino:armv7hf'
Installing (downloading, building) binaries...
Downloading conan_sources.tgz
sqlite3/3.35.4: Configuring sources in /builds/sqlite-build/conanCache/.conan/data/sqlite3/3.35.4/_/_/source
Unzipping 9.5MB, this can take a while
sqlite3/3.35.4: Copying sources to build folder
sqlite3/3.35.4: Building your package in /builds/sqlite-build/conanCache/.conan/data/sqlite3/3.35.4/_/_/build/70d57bfbf494755d82c86f4094add5b78d39e5d7
sqlite3/3.35.4: Generator cmake created conanbuildinfo.cmake
sqlite3/3.35.4: Calling build()
-- The C compiler identification is GNU 7.3.1
System is unknown to cmake, create:
Platform/FreeRTOS to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Detecting C compiler ABI info
System is unknown to cmake, create:
Platform/FreeRTOS to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/st/stm32cubeide_1.4.0/plugins/com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.7-2018-q2-update.linux64_1.4.0.202007081208/tools/bin/arm-none-eabi-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Conan: called by CMake conan helper
-- Conan: called inside local cache
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan: Adjusting fPIC flag (OFF)
-- This project seems to be plain C, using 'GNU' compiler
-- Conan: Compiler GCC>=5, checking major version 7
-- Conan: Checking correct version: 7
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:
    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
-- Build files have been written to: /builds/sqlite-build/conanCache/.conan/data/sqlite3/3.35.4/_/_/build/70d57bfbf494755d82c86f4094add5b78d39e5d7
Scanning dependencies of target sqlite3
[ 50%] Building C object CMakeFiles/sqlite3.dir/source_subfolder/sqlite3.c.obj
/builds/sqlite-build/conanCache/.conan/data/sqlite3/3.35.4/_/_/build/70d57bfbf494755d82c86f4094add5b78d39e5d7/source_subfolder/sqlite3.c:33756:10: fatal error: sys/ioctl.h: No such file or directory
 #include <sys/ioctl.h>
          ^~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/sqlite3.dir/build.make:82: CMakeFiles/sqlite3.dir/source_subfolder/sqlite3.c.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:95: CMakeFiles/sqlite3.dir/all] Error 2
make: *** [Makefile:149: all] Error 2
```

That include file must be used by Unix systems only. If I use define `SQLITE_OS_OTHER=1`, sources were compiled nice

**Comment by SpaceIm** [comment]
And with 2 profiles?

File: recipes/sqlite3/all/conanfile.py
Comment (AJIOB) [reply recipes/sqlite3/all/conanfile.py]: Thanks, you are right. I changed that option to boolean (using default or custom SQLite VFS)

**Comment by AJIOB** [comment]
> 
> 
> And with 2 profiles?

I changed an option for adding the possibility to disable the default VFS layer.

File: recipes/sqlite3/all/conanfile.py
Comment (madebr) [inline recipes/sqlite3/all/conanfile.py:94]: I found the original wording confusing.
```suggestion
            raise ConanInvalidConfiguration("build_executable=True cannot be combined with enable_default_vfs=False")
```

File: recipes/sqlite3/all/test_package/conanfile.py
Comment (madebr) [inline recipes/sqlite3/all/test_package/conanfile.py:11]: Would it be possible to create an empty/dummy vfs?
See https://www.sqlite.org/c3ref/vfs.html
Or is this out of scope?

At the end, the only thing we care about is whether the library can be linked against and whether basic functionality is there.

Comment (AJIOB) [reply recipes/sqlite3/all/test_package/conanfile.py:11]: Nowadays that implementation is out of the scope.

You can link created library if you provide `sqlite3_os_init` and `sqlite3_os_end` function implementations ([that document](https://www.sqlite.org/custombuild.html), chapter 5.0)

That's useful for cross-compiled embedded devices, where you need to use custom FS API access instead of libc `fopen`, `fclose` and others.

File: recipes/sqlite3/all/conanfile.py
Comment (AJIOB) [reply recipes/sqlite3/all/conanfile.py]: Thanks, that was fixed

File: recipes/sqlite3/all/test_package/empty_vfs.c
Comment (madebr) [inline recipes/sqlite3/all/test_package/empty_vfs.c:29]: image

File: recipes/sqlite3/all/test_package/conanfile.py
Comment (madebr) [reply recipes/sqlite3/all/test_package/conanfile.py]: Great dummy vfs implementation you've written!
The only thing I'm worried about now is how it will behave when the layout of the vfs struct changes. I guess the sqlite developers guarantee some stability?

Comment (AJIOB) [reply recipes/sqlite3/all/test_package/conanfile.py]: SQLite developers use versions for that. I implemented version 2 instead of the latest 3.

Comment (madebr) [reply recipes/sqlite3/all/test_package/conanfile.py]: Superbe!
Then I'll wait on your EAP approval in #4 and for ci to finish to approve this change.

File: recipes/sqlite3/all/test_package/empty_vfs.c
Comment (AJIOB) [reply recipes/sqlite3/all/test_package/empty_vfs.c:29]: As you can see [here](https://sqlite.org/c3ref/vfs.html), `xRandomness()` should fill `nBytes` in `zOut`.

I think, that the default written value will be a nice random too, not you?

Comment (madebr) [reply recipes/sqlite3/all/test_package/empty_vfs.c:29]: Your implementation is completely fine!
I was making an obligatory xkcd reference.

**Review by prince-chrismc — DISMISSED** [review]

File: recipes/sqlite3/all/conanfile.py
Comment (uilianries) [inline recipes/sqlite3/all/conanfile.py:92]:
```suggestion
    def validate(self):
        if not self.options.enable_default_vfs and self.options.build_executable:
```

Comment (AJIOB) [reply recipes/sqlite3/all/conanfile.py:92]: Looks like `validate()` method is experimental feature: [see here](https://docs.conan.io/en/latest/reference/conanfile/methods.html#validate). That is why I don't want to use it.

**Review by madebr — APPROVED** [review]

File: recipes/sqlite3/all/conanfile.py
Comment (prince-chrismc) [reply recipes/sqlite3/all/conanfile.py]: It's important to note, CCI is the test platform for new Conan features so we want to use them.

Rest assured it's safe to use... there's a lot of it already deployed.

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**
